lxd/auth/oidc/oidc.go

```diff
@@ -257,53 +208,178 @@ func (o *Verifier) IsRequest(r *http.Request) bool {
+func (o *Verifier) getCookies(r *http.Request) (string, string, error) {
+	idTokenCookie, err := r.Cookie(CookieNameIDToken)
+	if err != nil && err != http.ErrNoCookie {
+		return "", "", fmt.Errorf("Failed to get ID token cookie from request: %w", err)
+	}
+
+	var idToken string
+	if idTokenCookie != nil {
+		err = o.secureCookie.Decode(CookieNameIDToken, idTokenCookie.Value, &idToken)
+		if err != nil {
+			return "", "", fmt.Errorf("Failed to decrypt ID token cookie: %w", err)
+		}
 	}
 
-	keySet := rp.NewRemoteKeySet(http.DefaultClient, discoveryConfig.JwksURI)
+	refreshTokenCookie, err := r.Cookie(CookieNameRefreshToken)
+	if err != nil && err != http.ErrNoCookie {
+		return "", "", fmt.Errorf("Failed to get refresh token cookie from request: %w", err)
+	}
 
-	return op.NewAccessTokenVerifier(issuer, keySet), nil
+	var refreshToken string
+	if refreshTokenCookie != nil {
+		err = o.secureCookie.Decode(CookieNameRefreshToken, refreshTokenCookie.Value, &refreshToken)
+		if err != nil {
+			return "", "", fmt.Errorf("Failed to decrypt refresh token cookie: %w", err)
+		}
+	}
+
+	return idToken, refreshToken, nil
+}
+
+// setCookies encrypts the ID and refresh tokens and sets them in the HTTP response. Cookies are only set if they are
+// non-empty. If delete is true, the values are set to empty strings and the cookie expiry is set to unix zero time.
+func (o *Verifier) setCookies(w http.ResponseWriter, idToken string, refreshToken string, delete bool) error {
+	if idToken != "" || delete {
+		encodedIDTokenCookie, err := o.secureCookie.Encode(CookieNameIDToken, idToken)
+		if err != nil {
+			return fmt.Errorf("Failed to encrypt ID token: %w", err)
+		}
+
+		idTokenCookie := http.Cookie{
+			Name:     CookieNameIDToken,
+			Value:    encodedIDTokenCookie,
+			Path:     "/",
+			Secure:   true,
+			HttpOnly: true,
+			SameSite: http.SameSiteStrictMode,
+		}
+
+		if delete {
+			idTokenCookie.Value = ""
+			idTokenCookie.Expires = time.Unix(0, 0)
+		}
+
+		http.SetCookie(w, &idTokenCookie)
+	}
+
+	if refreshToken != "" || delete {
+		encodedRefreshToken, err := o.secureCookie.Encode(CookieNameRefreshToken, refreshToken)
+		if err != nil {
+			return fmt.Errorf("Failed to encrypt refresh token: %w", err)
+		}
+
+		refreshTokenCookie := http.Cookie{
+			Name:     CookieNameRefreshToken,
+			Value:    encodedRefreshToken,
+			Path:     "/",
+			Secure:   true,
+			HttpOnly: true,
+			SameSite: http.SameSiteStrictMode,
+		}
+
+		if delete {
+			refreshTokenCookie.Value = ""
+			refreshTokenCookie.Expires = time.Unix(0, 0)
+		}
+
+		http.SetCookie(w, &refreshTokenCookie)
+	}
+
+	return nil
 }
 
 // NewVerifier returns a Verifier.
 func NewVerifier(issuer string, clientid string, audience string) (*Verifier, error) {
-	cookieKey, err := uuid.New().MarshalBinary()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create UUID: %w", err)
+	cookieHashKey := securecookie.GenerateRandomKey(32)
+	cookieBlockKey := securecookie.GenerateRandomKey(32)
```

Comment:
Not sure if lxd supports it, but this would not work in a clustered/distributed setup. AFAICT, each server would have its own key and they wouldn't be able to verify cookies that were signed by the other servers. Also all cookies will be invalidated upon restart, which would mean that all users will be logged out (not sure if you care about this).

Reply:
@nsklikas a very good point indeed.

Reply:
Hmm yeah, this is fine if you open the browser on a single cluster member but will fail if there is a load balancer in front...

I wonder if we can derive these keys from the cluster private key somehow.

Reply:
You could potentially use the cluster key to derive a random key.
That is how the API's certificate is the same across cluster members.

Reply:
I'm now using a SHA512 (64 bytes) hash of the server private key (network cert private key in a cluster) for the cookie hash key and a SHA256 (32 bytes) hash for the cookie block key. This will be the same on all cluster members.

Reply:
It's now using HDKF to derive keys from the server private key. I haven't used a salt because each cluster member needs the same value.